nav2_costmap_2d/src/layered_costmap.cpp

```diff
@@ -274,10 +274,12 @@ bool LayeredCostmap::isCurrent()
 
 void LayeredCostmap::setFootprint(const std::vector<geometry_msgs::msg::Point> & footprint_spec)
 {
-  footprint_ = footprint_spec;
-  nav2_costmap_2d::calculateMinAndMaxDistances(
-    footprint_spec,
-    inscribed_radius_, circumscribed_radius_);
+  auto min_max = nav2_costmap_2d::calculateMinAndMaxDistances(footprint_spec);
```

Comment:
Ok, let me sorry to make you have to do this another time, but let me just suggest a different way here. I really don't love the idea of costmap having any more mutexes than it already has, this could easily create unforeseen problems because of the number of other mutexes at play. I don't know that it _will_ cause any issues, but I will say that the relationship with all the threads, costmap layers, and dynamic footprints are too complex for me to want to only rely on my own 15-minutes-of-code-reviewing to guarantee that this doesn't introduce some rare deadlock condition. 

- Keep this structure where `calculateMinAndMaxDistances` returns the pair 
- Make  circumscribed_radius_, inscribed_radius_ atomics 
- Use atomic set/get to promise they're set changed safely here (no need for the scoped lock then) 
- Perhaps footprint an atomic too?

Reply:
Atomics could fix `circumscribed_radius_`, `inscribed_radius_`, but `footprint_` is an array of points (`std::vector<geometry_msgs::msg::Point>`) which can't easily be atomic.
A mutex seems like the right solution here.
If you're dead set against it, maybe getFootprint() could return a shared_ptr?

Reply:
I made `footprint_` a shared_ptr internally and kept the interface the same for now.
(I would consider returning a shared pointer and not having to copy the array.)
This should handle the worst of the threading problems.
There is still a possibility that a client could do something like grab a inscribed_radius from one update and a circumscribed_radius from a different one, but there is no way that those values would be simply garbage like before.